### ISSUE TITLE
onnxruntime-web bug : 'return f4' is fast as normal, but 'return f1,f2,f3,f4' is very slow in wasm and webgl mode

### ISSUE THREAD

**Issue by yufengyao-lingoace** [microsoft/onnxruntime#12609]
(1)slow when return f1,f2,f3,f4
def forward(self, x):
        f1 = self.b1(x)
        f2 = self.b2(f1)  
        f3 = self.b3(f2) 
        f4 = self.b4(f3)
        return f1,f2,f3,f4
when I return f1,f2,f3,f4 ， the onnx used in chrome is very slow.  (ort.min.js using wasm or webgl)

(2) fast when return f4
def forward(self, x):
        f1 = self.b1(x)
        f2 = self.b2(f1)  
        f3 = self.b3(f2) 
        f4 = self.b4(f3)
        return f4
when I only return f4 ， the onnx used in chrome is very fast. (ort.min.js using wasm or webgl)
but in normal , the two way should be the same time
I also find that 
def forward(self, x):
        f1 = self.b1(x)
        f2 = self.b2(f1) 
        return f2
is faster than
def forward(self, x):
        f1 = self.b1(x)
        return f1
So anyone can tell me why?

**Comment by yufengyao-lingoace** [comment]
Anyone can help me?

**Comment by shalvamist** [comment]
Hey yufengyao,

Thanks for the details, can you please share a bit more info - 
1. The code looks like it's python - how are you deploying it to the browser ?
2. Do you mind sharing the model so we could reproduce the issue and debug it ?
3. Can you share the env set up and onnxruntime version used to produce this issue ?

**Comment by yufengyao-lingoace** [comment]
> 
Hello,
       Thanks to you reply. It is easy to reproduce the issue. I create the model with pytorch, and then exported an onnx file, then use onnxruntime-web to run the onnx model. The pytorch code is as follow:

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone = torchvision.models.mobilenet_v3_large()
        del self.backbone.avgpool
        del self.backbone.classifier
        self.b1 = nn.Sequential(self.backbone.features[0],
                                self.backbone.features[1])
        self.b2 = nn.Sequential(self.backbone.features[2],
                                self.backbone.features[3])
        self.b3 = nn.Sequential(self.backbone.features[4],
                                self.backbone.features[5], self.backbone.features[6])
        self.b4 = nn.Sequential(self.backbone.features[7],
                                self.backbone.features[8], self.backbone.features[9], self.backbone.features[10],
                                self.backbone.features[11], self.backbone.features[12], self.backbone.features[13],
                                self.backbone.features[14], self.backbone.features[15], self.backbone.features[16])

    def forward(self, x):
        f1 = self.b1(x)
        f2 = self.b2(f1)  
        f3 = self.b3(f2) 
        f4 = self.b4(f3)
        return f1,f2,f3,f4

**Comment by shalvamist** [comment]
Hey,

Thanks for the update and details. Looks like I have enough info to start investigating the issue.
I will try to reproduce the issue on my end and I will keep you posted with the results.

**Comment by yufengyao-lingoace** [comment]
Thanks for your reply, below is my  javascript code. The main function is detect().:

<!DOCTYPE html>
<html>
<meta http-equiv=Content-Type content="text/html; charset=utf-8">

<head>
    <title>test</title>
    <style>
        body {
            background-color: rgb(40, 40, 40);
            padding-top: 100px;
        }

        #main_canvas {
            display: block;
            margin: 0 auto;
            border: 1px solid red;
        }

        #btn-div {
            padding-top: 10px;
            width: 170px;
            display: block;
            margin: 0 auto;
        }
    </style>
</head>

<body>
    <video width="640" height="480" hidden></video>
    <canvas width="320" height="240" hidden></canvas>
    <canvas id="main_canvas" width="320" height="240"></canvas>
    <div id="btn-div">
        <button id="start" style="width:80px;height:25px;">open</button>
        <button id="snap" style="width:80px;height:25px;" hidden>clip</button>
        <button id="close" style="width:80px;height:25px;">close</button>
    </div>

    <script src="https://cdn.bootcdn.net/ajax/libs/onnxruntime-web/1.12.1/ort.js"></script>
    
    <script src="js/opencv.js"></script>
    <script>
        var canvas = document.getElementsByTagName('canvas')[0],
            canvas1 = document.getElementsByTagName('canvas')[1],
            context = canvas.getContext('2d'),
            video = document.getElementsByTagName("video")[0],
            snap = document.getElementById("snap"),
            close = document.getElementById("close"),
            start = document.getElementById("start"),
            MediaStreamTrack,
            session_face,
            session_facepoints98,
            cv2;

        var scaled_size = [320, 240];

        var init_model = async function () {
            options = {
                executionProviders: ['wasm'],
                // executionProviders: ['webgl'],

            };
            session = await ort.InferenceSession.create('./weights/test.onnx', options); 
            cv2 = await cv; //加载opencv
        }


        var detect = async function () {
            let imgData = context.getImageData(0, 0, 320, 240);
            var frame = cv2.matFromImageData(imgData);

            let dsize = new cv2.Size(320, 240);
            let frame_rgba = new cv2.Mat();

            cv2.resize(frame, frame_rgba, dsize, 0, 0, cv2.INTER_CUBIC);
            frame.delete();
            let frame_rgb = new cv2.Mat();
            cv2.cvtColor(frame_rgba, frame_rgb, cv2.COLOR_RGBA2RGB); 
            let data = frame_rgb.data;
            let new_Data = new Float32Array(3 * scaled_size[0] * scaled_size[1]);
            let index = 0;
            for (let i = 0; i < data.length; i += 3) { 
                new_Data[index] = data[i] / 255;
                index++;
            }
            for (let i = 1; i < data.length; i += 3) {
                new_Data[index] = data[i] / 255;
                index++;
            }
            for (let i = 2; i < data.length; i += 3) {
                new_Data[index] = data[i] / 255;
                index++;
            }

            let input = new ort.Tensor('float32', new_Data, [1, 3, scaled_size[1], scaled_size[0]]);

            let feeds = {
                "input": input,
            }
            var old_date = new Date();
            const results = await session.run(feeds); 
            var new_date = new Date();
            let fps = (1000 / (new_date - old_date)).toFixed(4);
            cv2.putText(frame_rgba, "fps: " + fps.toString(), new cv2.Point(10, 20), cv2.FONT_HERSHEY_PLAIN, 1, new cv2.Scalar(255, 0, 0, 255), 2);
            cv2.imshow(canvas1, frame_rgba);
            frame_rgb.delete();
            frame_back.delete();
            frame_rgba.delete();
        }


        window.onload = function () {
            init_model(); 

            start.addEventListener('click', function () {
                if (navigator.mediaDevices && navigator.mediaDevices.getUserMedia) {
                    navigator.mediaDevices.getUserMedia({
                        video: true,
                        audio: false
                    }).then(function (stream) {
                        MediaStreamTrack = typeof stream.stop === 'function' ? stream : stream.getTracks()[1];
                        video.srcObject = stream;
                        video.play();

                        setInterval(function () {
                            try {
                                context.drawImage(video, 0, 0, 640, 480, 0, 0, 320, 240);
                                detect();
                            }
                            catch (ex) {
                                console.log(ex);
                            }
                        }, 10);
                    }).catch(function (err) {
                        console.log(err);
                    });



                }
            });
            close.addEventListener('click', function () {
                MediaStreamTrack && MediaStreamTrack.stop();
            });
        }

    </script>
</body>

</html>

**Comment by yufengyao-lingoace** [comment]
Sorry, I think you have some problem with the onnx.export function.  The env is torch==1.10.1,  torchvision==0.11.2, python==3.8.8. Looking forwrad to your reply, thanks very much.

**Comment by yufengyao-lingoace** [comment]
I tried your code, and found a difference:

the right code is:
def forward(self, x):
    f1 = self.b1(x)

but yours is:
def forward(self, x):
    f1 = self.backbone(x)

just change self.backbone to self.b1 and then can export the onnx file. Thanks very much!

**Comment by shalvamist** [comment]
Hey,

Thanks for the response - I was able to generate the two models and test them with random inputs.
Here is the code - 
"
<script src="https://cdn.jsdelivr.net/npm/onnxruntime-web/dist/ort.min.js"></script>
<script>
   // use an async context to call onnxruntime functions.
   async function main() {
        try{
            document.write(`init model`);
            session = await ort.InferenceSession.create('./test_fast.onnx'); 
            console.log(session);

            // prepare inputs. a tensor need its corresponding TypedArray as data
            var scaled_size = [360, 240];
            let fps=0;
            for (let i = 0; i < 50; i++) {

                let new_Data = new Float32Array(3 * scaled_size[0] * scaled_size[1],() => Math.floor(Math.random() * 256));
                
                let input = new ort.Tensor('float32', new_Data, [1, 3, scaled_size[1], scaled_size[0]]);

                let feeds = {
                    "input": input,
                }
                // prepare feeds. use model input names as keys.

                // feed inputs and run
                var old_date = new Date();
                const results = await session.run(feeds); 
                var new_date = new Date();

                fps = fps+(new_date - old_date);

            }
            
            document.write(`FPS: ${(1000 / (fps/50)).toFixed(4)}`);

        } catch (e) {

            console.log(e);
            document.write(`failed to inference ONNX model: ${e}.`);
        }
    }
    main();
</script>
"

The two versions of the models (see attached) were able to get about the same amount of FPS for a 50 inference run - ~17 FPS. I couldn't see any major difference between them. Can you please try this on your end ?

**Comment by shalvamist** [comment]
<script src="https://cdn.jsdelivr.net/npm/onnxruntime-web/dist/ort.min.js"></script>
<script>
   // use an async context to call onnxruntime functions.
   async function main() {
        try{
            document.write(`init model`);
            session = await ort.InferenceSession.create('./test_fast.onnx'); 
            console.log(session);

            // prepare inputs. a tensor need its corresponding TypedArray as data
            var scaled_size = [360, 240];
            let fps=0;
            for (let i = 0; i < 50; i++) {

                let new_Data = new Float32Array(3 * scaled_size[0] * scaled_size[1],() => Math.floor(Math.random() * 256));
                
                let input = new ort.Tensor('float32', new_Data, [1, 3, scaled_size[1], scaled_size[0]]);

                let feeds = {
                    "input": input,
                }
                // prepare feeds. use model input names as keys.

                // feed inputs and run
                var old_date = new Date();
                const results = await session.run(feeds); 
                var new_date = new Date();

                fps = fps+(new_date - old_date);

            }
            
            document.write(`FPS: ${(1000 / (fps/50)).toFixed(4)}`);

        } catch (e) {

            console.log(e);
            document.write(`failed to inference ONNX model: ${e}.`);
        }
    }
    main();
</script>

**Comment by yufengyao-lingoace** [comment]
Hi, Thank you for you test. I tried my code again, and when use **wasm** mode,  I get 17fps for the tow models as the same as you. But when I use **webgl** mode,  the model(return f1,f2,f3,f4) got only 8 fps, and the model(return f1) got 30fps. It seems that this problem is only shown in webgl mode.  Return f4 is four times faster than return f1,f2,f3,f4. Please try this code again. Thanks very much!
options = {
                executionProviders: ['webgl'],
            };
            session = await ort.InferenceSession.create('./weights/test-fast.onnx', options);

**Comment by shalvamist** [comment]
Hi,

Sorry for a long wait on this issue. We had a few discussions on the topic and it looks like the performance degradation is due to the data transfer from GPU <-> CPU. This also coincides with the analysis we performed, we found that the more outputs are defined the more performance degradation we see. There are a few ways we can mitigate this performance loss but it will require some effort and most likely an output stacking feature to be implemented. 

As of now we don't have this feature planned on our road map but we took note and once resources free-up (or demand for this feature will increased) we'll revisit the implementation plan.

Hope this answer is suitable for you as of now. Please let me know if you have any comments.

Thanks

**Comment by yufengyao-lingoace** [comment]
Thanks very much. With the problem found out, I will try a stacking feature in my model. Thanks again. With besh wishes.

**Comment by liuyingbin123** [comment]
same problem,i have 5 outputs for my model,wasm is much faster than webgl
